src/CostFunction3D.cpp

```diff
@@ -2499,43 +2509,42 @@ void CostFunction3D::FFtransform(const real* Astate, real* Cstate)
 void CostFunction3D::calcHmatrix()
 {
   int n;
-  integer hi,m,mi;
+  uint64_t hi,m,mi;  // uint64_t
+  uint64_t nnz;
   real i,j,k;
   int ii,jj,kk,d,var;
-  integer cIndex,wgt_index;
+  uint64_t cIndex,wgt_index; // uint64_t
   int iNode,jNode,kNode;
   int iiNode,jjNode,kkNode;
   int iis,iie,jjs,jje,kks,kke;
-  int *Hlength;
-  integer *mTmp, *mIncr;
-  integer dst;
 
   real ibasis,jbasis,kbasis;
   real weight;
 
   GPTLstart("CostFunction3D::calcHmatrix");
 
-  std::cout << "Build H transform matrix...\n";
+  std::cout << "Build H transform matrix..." << std::endl;
   std::cout << "calcHmatrix: Grid dimensions: (" << iDim << ", " << jDim << ", " << kDim << ")" << std::endl;
 
   //GPTLstart("CostFunction3D::calcHmatrix:allocate");
 
-  Hlength = new int[mObs];
-  mPtr = new integer[nState+1];
-  IH   = new integer [mObs+1];
+  IH   = new uint64_t [mObs+1];    // uint64_t
+  IHt  = new uint64_t [nState+1];  // uint64_t
 
-  //GPTLstart("CostFunction3D::calcHmatrix:nonzeros");
+  nnz=0;
+  //JMD variable-interleave
+  //GPTLstart("CostFunction3D::calcHmatrix:nnz");
   // Determine the number of non-zeros in H
 	//#pragma omp parallel for private(m,mi,hi,i,j,k,ii,iis,iie,jj,jjs,jje,kk,kks,kke,ibasis,jbasis,kbasis,iiNode,jjNode,kkNode,iNode,jNode,kNode,var,d,wgt_index,weight,cIndex) //[8.1]
-  #pragma acc parallel loop vector gang vector_length(32) copyout(Hlength[0:mObs]) private(m,mi,hi,i,j,k,ii,iis,iie,jj,jjs,jje,kk,kks,kke,ibasis,jbasis,kbasis,iiNode,jjNode,kkNode,iNode,jNode,kNode,var,d,wgt_index,weight,cIndex)
+  #pragma acc parallel loop vector gang vector_length(32) private(m,mi,hi,iis,iie,jjs,jje,kks,kke,ibasis,jbasis,kbasis,iiNode,jjNode,kkNode,iNode,jNode,kNode,var,d,wgt_index,weight,cIndex) reduction(+:nnz)
```

Comment:
I do not think we need the long list of `private` clause here since they are all scalar variables.

Reply:
@sjsprecious I agree that this directive needs to be cleaned up significantly.  I was planning on doing the cleanup in a later PR as a general cleanup is not directly related to the H^t changes.

Reply:
That makes sense.